### DISCUSSION TITLE
Question about converge failed

### DISCUSSION THREAD

**Discussion by xiaoheimai95** [CURENT/andes#422] [Q&A] [answered]
Dear Dr. Cui and Dr. Wang,
I recently have been using your platform to generate some transient unstable data (power rotor angle instability and voltage instability) based on the IEEE39 with renewable outputs system (talked with you in another discussion). However, the "Time step reduced to zero. Convergence is not likely" problem often occurs in the calculation process. What's more, the system is a little extreme in the calculation process. Either the calculation doesn't converge, or it's perfectly stable. Could you please give me some suggestions on how to modify it?

**Comment by jinningwang** [comment]
Hello xiaoheimai95,

One clarification that I haven't received the PhD yet, so call me Jinning is fine, ^-^.
The first question, "Time step reduced to zero. Convergence is not likely" is usually caused by that some variables might reach the limit. The limit here can be the numeric convergence limit, and this sometimes can be caused by the ``Limiter``. For this issue, you can do the eigenvalue analysis ``EIG`` to see the devices that caused the instability and the associated variables.
The second question, it is expected for a system to return stable after disturbance. If not we'll say the system is not robust enough or the disturbance is too large. Here I don't have more comments on "perfectly stable". It might be reasonable to regard the diverge scenario as unstable. There are some approaches to improve the system stability, such as a stronger Exciter, a larger capacity Slack generator, or a reduced electrical distance.

@cuihantao Feel free to correct my answer if I misunderstand anything.

Regards,
Jinning

**Reply by xiaoheimai95** [reply]
Hello Jinning，
I think maybe my expression is not accurate enough. In fact, I want to find out some unstable cases for the research of control strategy. However, I found that in the simulation process, either the non-convergence mentioned above or the system could return to stable situation at the end. And actually I want to have more convergent but unstable dataset.

Best wishes

**Reply by jinningwang** [reply]
To my understanding, unstable is one major reason that the numeric solution does not converge. As for that you desire unstable but converged case, I don't have better comments for now.

Regards,
Jinning

**Reply by xiaoheimai95** [reply]
Thanks for answering!

**Comment by cuihantao** [comment, marked as answer]
Hi @xiaoheimai95 , thanks for reporting. Can you upload a case a minimal code example to reproduce the issue?

Nonconvergence is usually due to no power flow solution, but any example case can help to identify potential bugs.

**Reply by xiaoheimai95** [reply]
Here is the code:
```
case_fault = andes.get_case("ieee39_wind_toggle.xlsx")
ss = andes.load(case_fault, default_config=True, setup=False, no_output=True)
# modify bus
pq_buses = ss.PQ.bus.v
pq_shed_idx = ss.PQ.find_idx(keys='bus', values=pq_buses)
pq_p = np.array(ss.PQ.p0.v)
pq_q = np.array(ss.PQ.q0.v)
pq_p_new = pq_p * 1.35
pq_q_new = pq_q * 1.35
ss.PQ.alter("p0", pq_shed_idx, pq_p_new)
ss.PQ.alter("q0", pq_shed_idx, pq_q_new)
# modify Toggle
ss.Toggle.alter('dev', 'Toggler_1',  'Line_46')
ss.Toggle.alter('dev', 'Toggler_2',  'Line_46')
ss.Toggle.alter('t', 'Toggler_2', 1.3)
s.setup()
ss.PQ.config.p2p = 1
ss.PQ.config.q2q = 1
ss.PQ.config.p2z = 0
ss.PQ.config.q2z = 0
ss.PQ.config.p2i = 0
ss.PQ.config.q2i = 0
ss.PQ.config.pq2z = 0
ss.REECA1.config.allow_adjust = 0
ss.REECA1.set(src='Ipmin', idx='REECA1_10', attr='v', value=-15)
ss.WTTQA1.set(src='Temin', idx='WTTQA1_10', attr='v', value=-20)
ss.PFlow.run()
ss.TDS.config.criteria = 0
ss.TDS.config.tf = 10
ss.TDS.run()
```
Thanks for your help! Best wishes!
[ieee39_wind_toggle.xlsx](https://github.com/CURENT/andes/files/11007745/ieee39_wind_toggle.xlsx)

**Reply by cuihantao** [reply]
Thanks but I couldn't get `alter` to run before setting up the system. Mind to double check?

**Reply by cuihantao** [reply]
Some points for you to consider:

```python
import andes
import numpy as np

andes.config_logger(20)

case_fault = "ieee39_wind_toggle.xlsx"  # <- fault time of 0.3 sec is very long
ss = andes.load(case_fault, default_config=True, setup=False, no_output=True)

# modify bus
pq_buses = ss.PQ.bus.v
pq_shed_idx = ss.PQ.find_idx(keys='bus', values=pq_buses)
pq_p = np.array(ss.PQ.p0.v)
pq_q = np.array(ss.PQ.q0.v)

pq_p_new = pq_p * 1.35
pq_q_new = pq_q * 1.35
for idx, pn, qn in zip(pq_shed_idx, pq_p_new, pq_q_new): # <- changed
    ss.PQ.alter("p0", idx, pn)
    ss.PQ.alter("q0", idx, qn)

# modify Toggle
ss.Toggle.alter('dev', 'Toggler_1',  'Line_46')
ss.Toggle.alter('dev', 'Toggler_2',  'Line_46')
ss.Toggle.alter('t', 'Toggler_2', 1.3)
ss.setup()

ss.PQ.config.p2p = 0
ss.PQ.config.q2q = 0
ss.PQ.config.p2z = 1  # <- changed load model
ss.PQ.config.q2z = 1
ss.PQ.config.p2i = 0
ss.PQ.config.q2i = 0
ss.PQ.config.pq2z = 0

ss.REECA1.config.allow_adjust = 1  # <- why not? Temax of WTTQA1 is less than initial value. See output info.
ss.REECA1.set(src='Ipmin', idx='REECA1_10', attr='v', value=-15)
ss.WTTQA1.set(src='Temin', idx='WTTQA1_10', attr='v', value=0)  # <- torque cannot be negative
ss.PFlow.run()
ss.TDS.config.criteria = 0
ss.TDS.config.tf = 5
ss.TDS.run()
```

**Reply by xiaoheimai95** [reply]
Thanks! I will recheck this case.

**Comment by cuihantao** [comment]
Also see https://github.com/CURENT/andes/issues/424
